### PR TITLE
8330158: C2: Loop strip mining uses ABS with min int

### PR DESCRIPTION
**Pull request by rwestrel**
This fixes 3 calls to ABS with a min int argument. I think all of them
are harmless:

- in `PhaseIdealLoop::exact_limit()`, I removed the call to ABS. The
  check is for a stride of 1 or -1.
  
- in `OuterStripMinedLoopNode::adjust_strip_mined_loop()`, for the
  computation of `scaled_iters_long`, the stride is passed to `ABS()`
  and then implicitly casted to long. I now cast the stride to long
  before `ABS()`. For a min int stride, `LoopStripMiningIter * stride`
  overflows the int range for all values of `LoopStripMiningIter`
  except 0 or 1. Those values are handled early on in that method. So
  for a min in stride:
  ```
  (jlong)scaled_iters != scaled_iters_long
  ```
  is always true and the method returns early.
  
- in `OuterStripMinedLoopNode::adjust_strip_mined_loop()`, the
  computation of `short_scaled_iters` also calls `ABS()` with the
  stride as argument. But the result of that computation is only used
  if the test for:
  ```
  (jlong)scaled_iters != scaled_iters_long
  ```
  doesn't cause an early return of the method. I reordered statements
  so the `ABS()` calls happens after that test which will cause an early
  return if the stride is min int.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330158](https://bugs.openjdk.org/browse/JDK-8330158): C2: Loop strip mining uses ABS with min int (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [ecd940f3](https://git.openjdk.org/jdk/pull/18813/files/ecd940f38175bf68ba462170d6e7c887a4aac437)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [ecd940f3](https://git.openjdk.org/jdk/pull/18813/files/ecd940f38175bf68ba462170d6e7c887a4aac437)
 * [Martin Balao](https://openjdk.org/census#mbalao) (@martinuy - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18813/head:pull/18813` \
`$ git checkout pull/18813`

Update a local copy of the PR: \
`$ git checkout pull/18813` \
`$ git pull https://git.openjdk.org/jdk.git pull/18813/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18813`

View PR using the GUI difftool: \
`$ git pr show -t 18813`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18813.diff">https://git.openjdk.org/jdk/pull/18813.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18813#issuecomment-2060975944)

### CODE REVIEW

**Review by shipilev — APPROVED** [review]
All right, I agree with this reasoning. 

Have you tried running tests with #18751 applied?

**Comment by rwestrel** [comment]
Thanks for reviewing this.

> Have you tried running tests with #18751 applied?

I only ran the particular test that you mentioned in the bug.

**Comment by shipilev** [comment]
> I only ran the particular test that you mentioned in the bug.

All right, let me run tests with #18751 applied and see if we have any surprises.

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -2962,14 +2963,19 @@ void OuterStripMinedLoopNode::adjust_strip_mined_loop(PhaseIterGVN* igvn) {
   CountedLoopEndNode* inner_cle = inner_cl->loopexit();
 
   int stride = inner_cl->stride_con();
-  jlong scaled_iters_long = ((jlong)LoopStripMiningIter) * ABS(stride);
+  jlong scaled_iters_long = ((jlong)LoopStripMiningIter) * ABS((jlong)stride);
   int scaled_iters = (int)scaled_iters_long;
-  int short_scaled_iters = LoopStripMiningIterShortLoop* ABS(stride);
+  if ((jlong)scaled_iters != scaled_iters_long) {
+    // Remove outer loop and safepoint (too few iterations)
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:2969]: Please put more extended comment here. What you have in PR description would be nice.

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -2962,14 +2963,19 @@ void OuterStripMinedLoopNode::adjust_strip_mined_loop(PhaseIterGVN* igvn) {
...
+    remove_outer_loop_and_safepoint(igvn);
+    return;
+  }
+  int short_scaled_iters = LoopStripMiningIterShortLoop * ABS(stride);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:2973]: So stride is not MIN_INT here but the expression still can overflow. Should we use `jlong` for expression and `short_scaled_iters`? `iter_estimate` is `jlong`.

**Review by vnkozlov — COMMENTED** [review]
two comments

File: src/hotspot/share/opto/loopnode.cpp (hunk 2 above)
Comment (dean-long) [inline src/hotspot/share/opto/loopnode.cpp:2973]: Isn't it true that `stride` can be MIN_INT here, if LoopStripMiningIter == 1?

**Comment by dean-long** [comment]
There's also a less obvious use of an abs() idiom in LoopLimitNode::Ideal, when it does
```
2579      stride_p = -stride_con;
```
if stride_con is negative.  Does it make sense to fix that as part of this PR?

File: src/hotspot/share/opto/loopnode.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopnode.cpp:2972]: I added a comment. Does it look good to you?

File: src/hotspot/share/opto/loopnode.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopnode.cpp]: Good catch. I updated the patch as suggested.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by shipilev** [comment]
> All right, let me run tests with #18751 applied and see if we have any surprises.

I ran Maven CTW, Fuzzer tests, and the rest of OpenJDK jtregs with my ABS-checking patch applied, and there are no surprises. Looks good!

**Comment by shipilev** [comment]
@rwestrel, if you could integrate this, we can then go forward with #18751. Thanks!

**Review by martinuy — APPROVED** [review]
Looks good to me.

**Comment by rwestrel** [comment]
@shipilev @dean-long @vnkozlov @martinuy thanks for the reviews

**Pull request closed without merging**